-- Test cases for keyword arguments.


[case testTypeErrorInKeywordArgument]
void f(object o): pass
f(o=None()) # E: None not callable

[case testSimpleKeywordArgument]
void f(A a): pass
f(a=A())
f(a=object()) # E: Argument 1 to "f" has incompatible type "object"
class A: pass

[case testTwoKeywordArgumentsNotInOrder]
void f(A a, B b): pass
f(b=A(), a=A()) # E: Argument 1 to "f" has incompatible type "A"
f(b=B(), a=B()) # E: Argument 2 to "f" has incompatible type "B"
f(a=A(), b=B())
f(b=B(), a=A())
class A: pass
class B: pass

[case testOneOfSeveralOptionalKeywordArguments]
void f(A a=None, B b=None, C c=None): pass
f(a=A())
f(b=B())
f(c=C())
f(b=B(), c=C())
f(a=B()) # E: Argument 1 to "f" has incompatible type "B"
f(b=A()) # E: Argument 1 to "f" has incompatible type "A"
f(c=B()) # E: Argument 1 to "f" has incompatible type "B"
f(b=B(), c=A()) # E: Argument 2 to "f" has incompatible type "A"
class A: pass
class B: pass
class C: pass

[case testBothPositionalAndKeywordArguments]
void f(A a, B b): pass
f(A(), b=A()) # E: Argument 2 to "f" has incompatible type "A"
f(A(), b=B())
class A: pass
class B: pass

[case testContextSensitiveTypeInferenceForKeywordArg]
void f(A a, A[] b): pass
f(b=[], a=A())
class A: pass
[builtins fixtures/list.py]

[case testGivingSameKeywordArgumentTwice]
void f(A a, B b): pass
f(a=A(), b=B(), a=A()) # E: "f" gets multiple values for keyword argument "a"
f(b=B(), a=A(), b=B()) # E: "f" gets multiple values for keyword argument "b"
class A: pass
class B: pass

[case testGivingArgumentAsPositionalAndKeywordArg]
void f(A a, B b=None): pass
f(A(), a=A()) # E: "f" gets multiple values for keyword argument "a"
class A: pass
class B: pass

[case testGivingArgumentAsPositionalAndKeywordArg2]
void f(A a=None, B b=None): pass
f(A(), a=A()) # E: "f" gets multiple values for keyword argument "a"
class A: pass
class B: pass

[case testInvalidKeywordArgument]
void f(A a): pass
f(b=object()) # E: Unexpected keyword argument "b" for "f"
class A: pass

[case testKeywordArgumentsWithDynamicallyTypedCallable]
any f
f(x=f(), z=None()) # E: None not callable
f(f, zz=None()) # E: None not callable
f(x=None)

[case testDuplicateKeywordArgumentWithDynamicallyTypedCallable]
any f
# Postpone checking to runtime
f(x=None, x=None)

[case testKeywordArgumentWithFunctionObject]
func<A, B, void> f
f(a=A(), b=B())
f(A(), b=B())
class A: pass
class B: pass
[out]
main, line 2: Unexpected keyword argument "a"
main, line 2: Unexpected keyword argument "b"
main, line 3: Unexpected keyword argument "b"

[case testKeywordOnlyArguments]
void f(A a, *, B b=None): pass
f(A(), b=B())
f(b=B(), a=A())
f(A())
f(A(), B()) # E: Too many positional arguments for "f"
class A: pass
class B: pass

[case testBareAsteriskWithVarArgs]
void f(A *a, *, B b=None): pass
f()
f(A())
f(A(), A())
f(b=B())
f(A(), b=B())
f(A(), A(), b=B())
f(B())      # E: Argument 1 to "f" has incompatible type "B"
f(A(), B()) # E: Argument 2 to "f" has incompatible type "B"
f(b=A())    # E: Argument 1 to "f" has incompatible type "A"
class A: pass
class B: pass
[builtins fixtures/list.py]

[case testBareAsteriskWithBothCallerAndCalleeVarArgs]
void f(A *a, *, B b=None): pass
A[] a
f(*a)
f(A(), *a)
f(b=B())
f(*a, b=B())
f(A(), *a, b=B())
f(A(), B())   # E: Argument 2 to "f" has incompatible type "B"
f(A(), b=A()) # E: Argument 2 to "f" has incompatible type "A"
f(*a, b=A())  # E: Argument 2 to "f" has incompatible type "A"
class A: pass
class B: pass
[builtins fixtures/list.py]

[case testCallingDynamicallyTypedFunctionWithKeywordArgs]
def f(x, y=A()): pass
f(x=A(), y=A())
f(y=A(), x=A())
f(y=A())      # E: Too few arguments for "f"
f(A(), z=A()) # E: Unexpected keyword argument "z" for "f"
class A: pass

[case testKwargsArgumentInFunctionBody]
void f(A **kwargs):
    dict<str, A> d1 = kwargs
    dict<A, any> d2 = kwargs # E: Incompatible types in assignment
    dict<any, str> d3 = kwargs # E: Incompatible types in assignment
class A: pass
[builtins fixtures/dict.py]
[out]
main: In function "f":

[case testKwargsArgumentInFunctionBodyWithImplicitAny]
void f(**kwargs):
    dict<str, A> d1 = kwargs
    dict<str, str> d2 = kwargs
    dict<A, any> d3 = kwargs # E: Incompatible types in assignment
class A: pass
[builtins fixtures/dict.py]
[out]
main: In function "f":

[case testCallingFunctionThatAcceptsVarKwargs]
void f(A **kwargs): pass
f(x=A())
f(y=A(), z=A())
f(x=B()) # E: Argument 1 to "f" has incompatible type "B"
f(A())   # E: Too many arguments for "f"
# Perhaps a better message would be "Too many *positional* arguments..."
class A: pass
class B: pass
[builtins fixtures/dict.py]
