-- Test cases for function overloading


[case testTypeCheckOverloadedFunctionBody]
from typing import overload
@overload
def f(x: 'A'):
    x = B() # E: Incompatible types in assignment
    x = A()
@overload
def f(x: 'B'):
    x = A() # E: Incompatible types in assignment
    x = B()
class A: pass
class B: pass
[out]
main: In function "f":

[case testTypeCheckOverloadedMethodBody]
from typing import overload
class A:
    @overload
    def f(self, x: 'A'):
        x = B() # E: Incompatible types in assignment
        x = A()
    @overload
    def f(self, x: 'B'):
        x = A() # E: Incompatible types in assignment
        x = B()
class B: pass
[out]
main: In member "f" of class "A":

[case testCallToOverloadedFunction]
from typing import overload
f(C()) # E: No overload variant of "f" matches argument types
f(A())
f(B())

@overload
def f(x: 'A') -> None: pass
@overload
def f(x: 'B') -> None: pass

class A: pass
class B: pass
class C: pass

[case testOverloadedFunctionReturnValue]
from typing import Undefined, overload, disjointclass
a, b = Undefined, Undefined # type: (A, B)
b = f(a) # E: Incompatible types in assignment
a = f(b) # E: Incompatible types in assignment
a = f(a)
b = f(b)

@overload
def f(x: 'A') -> 'A': pass
@overload
def f(x: 'B') -> 'B': pass
class A: pass
@disjointclass(A)
class B: pass

[case testCallToOverloadedMethod]
from typing import overload
A().f(C()) # E: No overload variant of "f" of "A" matches argument types
A().f(A())
A().f(B())

class A:
  @overload
  def f(self, x: 'A') -> None: pass
  @overload
  def f(self, x: 'B') -> None: pass

class B: pass
class C: pass

[case testOverloadedMethodReturnValue]
from typing import Undefined, overload, disjointclass
a, b = Undefined, Undefined # type: (A, B)
b = a.f(a) # E: Incompatible types in assignment
a = a.f(b) # E: Incompatible types in assignment
a = a.f(a)
b = a.f(b)

class A:
  @overload
  def f(self, x: 'A') -> 'A': pass
  @overload
  def f(self, x: 'B') -> 'B': pass
@disjointclass(A)
class B: pass

[case testNonConsecutiveOverloads]
from typing import Any
def f() -> None: pass
def g() -> None: pass
def f(x: Any) -> None: pass # E: Name 'f' already defined (overload variants must be next to each other)

[case testOverloadsWithDifferentArgumentCounts]
from typing import Undefined, overload
a, b = Undefined, Undefined # type: (A, B)
a = f(a)
b = f(a) # E: Incompatible types in assignment
f(b)     # E: No overload variant of "f" matches argument types
b = f(b, a)
a = f(b, a) # E: Incompatible types in assignment
f(a, a)     # E: No overload variant of "f" matches argument types
f(b, b)     # E: No overload variant of "f" matches argument types

@overload
def f(x: 'A') -> 'A': pass
@overload
def f(x: 'B', y: 'A') -> 'B': pass
class A: pass
class B: pass

[case testGenericOverloadVariant]
from typing import Undefined, overload, typevar, Generic, disjointclass
t = typevar('t')
ab, ac, b, c = Undefined, Undefined, Undefined, Undefined # type: (A[B], A[C], B, C)
b = f(ab)
c = f(ac)
b = f(ac) # E: Argument 1 to "f" has incompatible type A[C]; expected A[B]
b = f(b)
c = f(b)  # E: Incompatible types in assignment
@overload
def f(x: 'A[t]') -> t: pass
@overload
def f(x: 'B') -> 'B': pass
class A(Generic[t]): pass
@disjointclass(A)
class B: pass
class C: pass

[case testOverloadedInit]
from typing import Undefined, overload
a, b = Undefined, Undefined # type: (A, B)
a = A(a)
a = A(b)
a = A(object()) # E: No overload variant of "A" matches argument types

class A:
  @overload
  def __init__(self, a: 'A') -> None: pass
  @overload
  def __init__(self, b: 'B') -> None: pass
class B: pass

[case testIntersectionTypeCompatibility]
from typing import Undefined, overload, Function
o = Undefined # type: object
a = Undefined # type: A

a = f # E: Incompatible types in assignment
o = f

@overload
def f(a: 'A') -> None: pass
@overload
def f(a: Function[[], None]) -> None: pass
class A: pass

[case testCompatibilityOfIntersectionTypeObjectWithStdType]
from typing import Undefined, overload
t, a = Undefined, Undefined # type: (type, A)

a = A # E: Incompatible types in assignment
t = A

class A:
    @overload
    def __init__(self, a: 'A') -> None: pass    
    @overload
    def __init__(self, a: 'B') -> None: pass
class B: pass

[case testOverloadedGetitem]
from typing import Undefined, overload
a, b = Undefined(int), Undefined(str)
a = A()[a]
b = A()[a] # E: Incompatible types in assignment
b = A()[b]
a = A()[b] # E: Incompatible types in assignment

class A:
    @overload
    def __getitem__(self, a: int) -> int: pass
    @overload
    def __getitem__(self, b: str) -> str: pass

[case testOverloadedGetitemWithGenerics]
from typing import Undefined, typevar, Generic, overload
t = typevar('t')
a, b, c = Undefined, Undefined, Undefined # type: (A, B, C[A])
a = c[a]
b = c[a] # E: Incompatible types in assignment
a = c[b]
b = c[b] # E: Incompatible types in assignment

class C(Generic[t]):
    @overload
    def __getitem__(self, a: 'A') -> t: pass
    @overload
    def __getitem__(self, b: 'B') -> t: pass
class A: pass
class B: pass

[case testImplementingOverloadedMethod]
from typing import overload
from abc import abstractmethod, ABCMeta

class I(metaclass=ABCMeta):
    @overload
    @abstractmethod
    def f(self) -> None: pass
    @overload
    @abstractmethod
    def f(self, a: 'A') -> None: pass
class A(I):
    @overload
    def f(self) -> None: pass
    @overload
    def f(self, a: 'A') -> None: pass

[case testOverloadWithFunctionType]
from typing import overload, Function
class A: pass
@overload
def f(x: A) -> None: pass
@overload
def f(x: Function[[], None]) -> None: pass

f(A())
[builtins fixtures/function.py]

[case testVarArgsOverload]
from typing import overload, Any
@overload
def f(x: 'A', *more: Any) -> 'A': pass
@overload
def f(x: 'B', *more: Any) -> 'A': pass
f(A())
f(A(), A, A)
f(B())
f(B(), B)
f(B(), B, B)
f(object()) # E: No overload variant of "f" matches argument types
class A: pass
class B: pass
[builtins fixtures/list.py]

[case testVarArgsOverload2]
from typing import overload
@overload
def f(x: 'A', *more: 'B') -> 'A': pass
@overload
def f(x: 'B', *more: 'A') -> 'A': pass
f(A(), B())
f(A(), B(), B())
f(A(), A(), B()) # E: No overload variant of "f" matches argument types
f(A(), B(), A()) # E: No overload variant of "f" matches argument types
class A: pass
class B: pass
[builtins fixtures/list.py]

[case testOverloadWithTypeObject]
from typing import overload
@overload
def f(a: 'A', t: type) -> None: pass
@overload
def f(a: 'B', t: type) -> None: pass
f(A(), B)
f(B(), A)
class A: pass
class B: pass
[builtins fixtures/function.py]

[case testOverloadedInitAndTypeObjectInOverload]
from typing import overload, Undefined, disjointclass
@overload
def f(t: type) -> 'A': pass
@overload
def f(t: 'A') -> 'B': pass
a, b = Undefined, Undefined # type: (A, B)
a = f(A)
b = f(a)
b = f(A) # E: Incompatible types in assignment
a = f(a) # E: Incompatible types in assignment
@disjointclass(type)
class A:
   @overload
   def __init__(self) -> None: pass
   @overload
   def __init__(self, a: 'A') -> None: pass
class B:
    pass

[case testOverlappingErasedSignatures]
from typing import overload, List
@overload
def f(a: List[int]) -> int: pass
@overload
def f(a: List[str]) -> int: pass
n = f(List[int]())
m = f(List[str]())
n = 1
m = 1
n = 'x' # E: Incompatible types in assignment
m = 'x' # E: Incompatible types in assignment
f(List[object]()) # E: Argument 1 to "f" has incompatible type List[object]; expected List[int]
[builtins fixtures/list.py]

[case testOverlappingOverloadSignatures]
from typing import overload
class A: pass
class B: pass
@overload
def f(x: A) -> A: pass # E: Overloaded function signatures 1 and 2 overlap with incompatible return types
@overload
def f(x: B) -> B: pass

[case testDisjointclassesAndOverload]
from typing import overload, disjointclass
class A: pass
@disjointclass(A)
class B: pass
@overload
def f(x: A) -> A: pass
@overload
def f(x: B) -> B: pass

[case testContravariantOverlappingOverloadSignatures]
from typing import overload
class A: pass
class B(A): pass
@overload
def f(x: A) -> A: pass
@overload
def f(x: B) -> B: pass # This is more specific than the first item, and thus
                       # will never be called.

[case testPartiallyCovariantOverlappingOverloadSignatures]
from typing import overload
class A: pass
class B(A): pass
@overload
def f(x: B) -> A: pass # E: Overloaded function signatures 1 and 2 overlap with incompatible return types
@overload
def f(x: A) -> B: pass

[case testPartiallyContravariantOverloadSignatures]
from typing import overload
class A: pass
class B(A): pass
@overload
def g(x: A) -> int: pass # Fine, since A us supertype of B.
@overload
def g(x: B) -> str: pass

[case testCovariantOverlappingOverloadSignatures]
from typing import overload
class A: pass
class B(A): pass
@overload
def g(x: B) -> B: pass
@overload
def g(x: A) -> A: pass

[case testCovariantOverlappingOverloadSignaturesWithSomeSameArgTypes]
from typing import overload
class A: pass
class B(A): pass
@overload
def g(x: int, y: B) -> B: pass
@overload
def g(x: int, y: A) -> A: pass

[case testCovariantOverlappingOverloadSignaturesWithAnyType]
from typing import Any, overload
@overload
def g(x: int) -> int: pass
@overload
def g(x: Any) -> Any: pass

[case testContravariantOverlappingOverloadSignaturesWithAnyType]
from typing import Any, overload
@overload
def g(x: Any) -> Any: pass # E: Overloaded function signatures 1 and 2 overlap with incompatible return types
@overload
def g(x: int) -> int: pass

[case testOverloadedLtAndGtMethods]
from typing import overload
class A:
    def __lt__(self, x: A) -> int: pass
    def __gt__(self, x: A) -> int: pass
class B:
    @overload
    def __lt__(self, x: B) -> int: pass
    @overload
    def __lt__(self, x: A) -> int: pass
    @overload
    def __gt__(self, x: B) -> int: pass
    @overload
    def __gt__(self, x: A) -> int: pass
A() < A()
A() < B()
B() < A()
B() < B()
A() < object() # E: Unsupported operand types for < ("A" and "object")
B() < object() # E: No overload variant of "__lt__" of "B" matches argument types

[case testOverloadedForwardMethodAndCallingReverseMethod]
from typing import overload
class A:
    @overload
    def __add__(self, x: 'A') -> int: pass
    @overload
    def __add__(self, x: int) -> int: pass
class B:
    def __radd__(self, x: A) -> int: pass
A() + A()
A() + 1
A() + B()
A() + '' # E: No overload variant of "__add__" of "A" matches argument types

