-- Nested list assignment
-- -----------------------------


[case testNestedListAssignment]
from typing import Undefined, List
a1, b1, c1 = Undefined, Undefined, Undefined # type: (A, B, C)
a2, b2, c2 = Undefined, Undefined, Undefined # type: (A, B, C)

a1, [b1, c1] = a2, [b2, c2]
a1, [a1, [b1, c1]] = a2, [a2, [b2, c2]]
a1, [a1, [a1, b1]] = a1, [a1, [a1, c1]]  # E: Incompatible types in assignment (expression has type "C", variable has type "B")

class A: pass
class B: pass
class C: pass
[builtins fixtures/list.py]
[out]

[case testNestedListAssignmenToTuple]
from typing import Undefined, List
a, b, c = Undefined, Undefined, Undefined # type: (A, B, C)

a, b = [a, b]
a, b = [a]  # E: Need 2 values to assign
a, b = [a, b, c]  # E: Too many values to assign

class A: pass
class B: pass
class C: pass
[builtins fixtures/list.py]
[out]

[case testListAssignmentFromTuple]
from typing import Undefined, List
a, b, c = Undefined, Undefined, Undefined # type: (A, B, C)
t = a, b

[a, b], c = t, c
[a, c], c = t, c  # E: Incompatible types in assignment (expression has type "B", variable has type "C")
[a, a, a], c = t, c  # E: Need more than 2 values to unpack (3 expected)
[a], c = t, c  # E: Too many values to unpack (1 expected, 2 provided)

class A: pass
class B: pass
class C: pass
[builtins fixtures/list.py]
[out]

[case testListAssignmentUnequalAmountToUnpack]
from typing import Undefined, List
a, b, c = Undefined, Undefined, Undefined # type: (A, B, C)

a, b = [a, b]
a, b = [a]  # E: Need 2 values to assign
a, b = [a, b, c]  # E: Too many values to assign

class A: pass
class B: pass
class C: pass
[builtins fixtures/list.py]
[out]